packages/core/src/pubsub/__tests__/message-bus.test.ts

```diff
@@ -74,30 +74,69 @@ test('unsubscribe', async () => {
 });
 
 describe('queryNetwork', () => {
-  const FAKE_CID = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu');
+  const FAKE_CID1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu');
+  const FAKE_CID2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu');
   const queryMessage = buildQueryMessage(FAKE_STREAM_ID);
-  const responseMessage: ResponseMessage = {
+  const responseMessage1: ResponseMessage = {
     typ: MsgType.RESPONSE,
     id: queryMessage.id,
-    tips: new Map().set(FAKE_STREAM_ID.toString(), FAKE_CID),
+    tips: new Map().set(FAKE_STREAM_ID.toString(), FAKE_CID1),
   };
-  const pubsub = (from([responseMessage]).pipe(delay(300)) as unknown) as Pubsub;
+  const responseMessage2: ResponseMessage = {
+    typ: MsgType.RESPONSE,
+    id: queryMessage.id,
+    tips: new Map().set(FAKE_STREAM_ID.toString(), FAKE_CID2),
+  };
+  const pubsub = (from([responseMessage1, responseMessage2]).pipe(delay(200)) as unknown) as Pubsub;
   pubsub.next = jest.fn();
 
-  test('return tip', async () => {
+  test('return tips', async () => {
+    const messageBus = new MessageBus(pubsub);
+    const responses = [];
+    const subscription = messageBus.queryNetwork(FAKE_STREAM_ID).subscribe((r) => {
+      responses.push(r);
+    });
+    await new Promise<void>((resolve) => subscription.add(resolve));
+    expect(pubsub.next).toBeCalledWith(queryMessage);
+    expect(responses).toEqual([FAKE_CID1, FAKE_CID2]);
+  });
+
+  test('delayed message', async () => {
+    const pubsub = (from([responseMessage1, responseMessage2]).pipe(
+      concatMap(async (r) => {
+        await new Promise((resolve) => setTimeout(resolve, MAX_RESPONSE_INTERVAL * 2));
+        return r;
+      }),
+    ) as unknown) as Pubsub;
+    pubsub.next = jest.fn();
     const messageBus = new MessageBus(pubsub);
-    const response = await messageBus.queryNetwork(FAKE_STREAM_ID).toPromise();
+    const responses = [];
+    const subscription = messageBus.queryNetwork(FAKE_STREAM_ID).subscribe((r) => {
+      responses.push(r);
+    });
+    await new Promise<void>((resolve) => subscription.add(resolve));
     expect(pubsub.next).toBeCalledWith(queryMessage);
-    expect(response).toEqual(FAKE_CID);
+    expect(responses).toEqual([FAKE_CID1]);
   });
 
-  test('timeout: return undefined', async () => {
+  test('late first message', async () => {
+    const pubsub = (new Observable<ResponseMessage>((subscriber) => {
+      setTimeout(() => {
+        subscriber.next(responseMessage1);
+        setTimeout(() => {
+          subscriber.next(responseMessage2);
+          subscriber.complete();
+        }, MAX_RESPONSE_INTERVAL * 2);
```

Comment:
So I believe the test above ('delayed message') is testing what happens when there's plenty of time in the outer 3 second syncTimeoutSeconds, but the time between the two responses is larger than the smaller 300ms inner MAX_RESPONSE_INTERVAL, so the second response doesn't get processed because it came in too late.

There's another case, which I *think* is what this test is trying to test? That case is where a response comes in close to the end of the outer 3 second syncTimeoutSeconds, and then another response comes in in *less* than the 300ms MAX_RESPONSE_INTERVAL, but by that point the 3 second syncTimeoutSeconds time limit is up, so even though the second response came in less than 300ms after the first, it doesn't get processed because we've exhausted our outer time limit.

Maybe that's not what this test was trying to test, but in any case I don't think this tests that since the second response comes in more than MAX_RESPONSE_INTERVAL after the first. Seems like this would be a good case to test, but I think maybe it would need to happen in state-manager.test.ts, since it's the state manager that enforces the outer 3 second syncTimeoutSeconds?

Hopefully I explained that right and it's clear what I'm referring to, please let me know if anything isn't clear

Reply:
I guess the 'stop after timeout' test in the state manager tests basically tests the case I'm talking about.  So then I guess I just don't understand what case this 'late first message' test here is trying to test?

Reply:
As stated in the name, it tests for the case when the first received message is late, that is arrives after the in-between timeout has passed from the start time. The timeout must only be counted between the messages, not between start time and the first message.

Reply:
Ah got it, that makes sense.  Thanks for clarifying.